include/picongpu/plugins/misc/rangeToWindow.hpp

```diff
@@ -0,0 +1,115 @@
+             *                      cropped to fit the data domain.
+             * @return Window description with local and global information.
+             */
+            template<uint32_t T_dim>
+            inline Window rangeToWindow(
+                pmacc::SubGrid<T_dim> const& dataDomain,
+                Window const& inputWindow,
+                std::string const& selectedRange)
+            {
+                const SubGrid<simDim>& subGrid = Environment<simDim>::get().SubGrid();
+
+                auto parsedSlice = pmacc::pluginSystem::toTimeSlice(selectedRange);
+
+                Window resultWindow = inputWindow;
+                // Parse the ranges only for dimensions the user is providing.
+                for(uint32_t d = 0; d < parsedSlice.size() && d < simDim; ++d)
+                {
+                    auto range = parsedSlice[d];
+                    uint32_t const windowOffset = inputWindow.globalDimensions.offset[d];
+                    uint32_t const windowSize = inputWindow.globalDimensions.size[d];
+                    uint32_t const windowEnd = windowOffset + windowSize;
+                    uint32_t const rangeBegin = range.values[0];
+                    // Crop the user given end range to the end of the input window.
+                    uint32_t const rangeEnd = std::min(range.values[1], windowEnd);
+
+                    // Crop resulting window begin to the end of the input window.
+                    uint32_t const newWindowBegin = std::min(rangeBegin + windowOffset, windowEnd);
+                    // Crop resulting window end to the end of the input window.
+                    uint32_t const newWindowEnd = std::min(rangeEnd + windowOffset, windowEnd);
+                    uint32_t const newWindowSize = newWindowEnd - newWindowBegin;
+
+                    resultWindow.globalDimensions.offset[d] = newWindowBegin;
+                    resultWindow.globalDimensions.size[d] = newWindowSize;
+
+                    /* Update offset from the local data domain with respect to the window begin.
+                     * In cases where the window begin is inside of the local domain of the device the offset is zero
+                     * and not negative.
+                     */
+                    resultWindow.localDimensions.offset[d]
+                        = std::max(0, subGrid.getLocalDomain().offset[d] - static_cast<int>(newWindowBegin));
+
+
+                    /* If the resulting window is before or after the local data domain the window size will be local
+                     * zero because the local domain is outside the window else the local domain is part of the window.
+                     */
+                    if(newWindowEnd <= subGrid.getLocalDomain().offset[d]
+                       || subGrid.getLocalDomain().size[d] + subGrid.getLocalDomain().offset[d] < newWindowBegin)
```

Comment:
Can't we remove this `if`, always do second branch and then crop to 0 if calculated size becomes negative?

Reply:
No this `if` is required because there is no need that the range to be spanned over all devices.
It is possible to define a range with a single cell somewhere in the middle. In this case there are GPUs before the window and after the window which are not participating in the data gathering process.